Source/Dafny/Translator.cs

```diff
@@ -94,6 +96,229 @@ public string FreshNumericId(string prefix = "")
     }
   }
 
+  class SingularDriver
+  {
+    static uint queryCount = 0;
+
+    uint tempVarCount = 0;
+    static string tempVarPrefix = "tpv_";
+
+    List<string> polys;
+    HashSet<string> variables;
+    Dictionary<string, string> opaqueExprs;
+
+    TextWriter writer;
+
+    string AllocateFreshTemp() {
+      var name = String.Format("{0}{1}", SingularDriver.tempVarPrefix, this.tempVarCount);
+      this.tempVarCount += 1;
+      this.variables.Add(name);
+      return name;
+    }
+
+    string OpacifySubExpr(Expression expr) {
+      if (expr is SeqSelectExpr) {
```

Comment:
C# will let you write:
```
if (expr is SeqSelectExpr seqSelExpr) {
  ...
```
i.e., you can combine the test with the naming of the new variable.  The name scoping is annoying though, each new variable needs to be unique, so you can't just call them all `e`, for example.

Reply:
Ah that's good to know. Thanks for the tip!